charmhelpers/contrib/network/ovs/__init__.py

```diff
@@ -303,3 +306,158 @@ def port_to_br(port):
+        more details.
+
+        NOTE(fnordahl): Technically the ``external-ids`` column is called
+        ``external_ids`` (with an underscore) and we rely on ``ovs-vsctl``'s
+        behaviour of transforming dashes to underscores for us [3] so we can
+        have a more pleasant data structure.
+
+        0: http://www.openvswitch.org/ovs-vswitchd.conf.db.5.pdf
+        1: https://tools.ietf.org/html/rfc7047
+        2: http://docs.openvswitch.org/en/latest/topics/integration/
+        3: https://github.com/openvswitch/ovs/blob/
+               20dac08fdcce4b7fda1d07add3b346aa9751cfbc/
+                   lib/db-ctl-base.c#L189-L215
+    :type data: Optional[Dict[str,Union[str,Dict[str,str]]]]
+    :param table: Name of table to operate on
+    :type table: str
+    :param entity: Name of entity to operate on
+    :type entity: str
+    :returns: '--' separated ``ovs-vsctl set`` commands
+    :rtype: Iterator[Tuple[str, str, str, str, str]]
+    """
+    for (k, v) in data.items():
+        if isinstance(v, dict):
+            entries = {
+                '{}:{}'.format(k, dk): dv for (dk, dv) in v.items()}
+        else:
+            entries = {k: v}
+        for (colk, colv) in entries.items():
+            yield ('--', 'set', table, entity, '{}={}'.format(colk, colv))
+
+
+def add_br(bridge, brdata=None, exclusive=False):
+    """Add bridge and optionally attach a external_id to bridge.
+
+    :param bridge: Name of bridge to create
+    :type bridge: str
+    :param brdata: Additional data to attach to bridge
+        The keys in the brdata dictionary map directly to column names in the
+        OpenvSwitch bridge table as defined in DB-SCHEMA [0] referenced in
+        RFC 7047 [1]
+
+        There are some established conventions for keys in the external-ids
+        column of various tables, consult the OVS Integration Guide [2] for
+        more details.
+
+        NOTE(fnordahl): Technically the ``external-ids`` column is called
+        ``external_ids`` (with an underscore) and we rely on ``ovs-vsctl``'s
+        behaviour of transforming dashes to underscores for us [3] so we can
+        have a more pleasant data structure.
+
+        0: http://www.openvswitch.org/ovs-vswitchd.conf.db.5.pdf
+        1: https://tools.ietf.org/html/rfc7047
+        2: http://docs.openvswitch.org/en/latest/topics/integration/
+        3: https://github.com/openvswitch/ovs/blob/
+               20dac08fdcce4b7fda1d07add3b346aa9751cfbc/
+                   lib/db-ctl-base.c#L189-L215
+    :type brdata: Optional[Dict[str,Union[str,Dict[str,str]]]]
+    :param exclusive: If True, raise exception if bridge exists
+    :type exclusive: bool
+    :raises: subprocess.CalledProcessError
+    """
+    cmd = ['ovs-vsctl', 'add-br', bridge]
+    if not exclusive:
+        cmd.append('--may-exist')
+    if brdata:
+        for setcmd in _dict_to_vsctl(brdata, 'bridge', bridge):
+            cmd.extend(setcmd)
+    utils._run(*cmd)
```

Comment:
This seems to be broken. In my case I call it like this:

```
            add_br(
                'br-int', brdata={
                    'external-ids': {
                        OVS_EXTERNAL_ID_KEY: OVS_EXTERNAL_ID_BRIDGE_VALUE
                    }
                }
            )

```

and it produces the following command:

```
ovs-vsctl add-br br-int --may-exist -- set bridge br-int external-ids:charm-neutron-gateway=managed
```

which exits with 1 and prints:

```
ovs-vsctl: 'add-br' command takes exactly 1 or 3 arguments
```

It looks like the former `add_bridge()` was producing a slightly different command that would look like `ovs-vsctl -- --may-exist add-br ...`

Reply:
Suggested fix: https://github.com/AurelienLourot/charm-neutron-gateway/commit/57e27117
Works for charm-neutron-gateway

Reply:
Yes, that got lost in translation apparently, ta.